File: court-booking-frontend/packages/types/src/payment.ts
import type { Booking } from "./booking";

export type OcrVerdict = "match" | "mismatch" | "unreadable";
export type ReviewVerdict = "approved" | "rejected";

/** Section 32 Part 7 -- verdicts computed once at submission time. */
export type NameMatchVerdict = "match" | "mismatch" | "unavailable";
export type TimeCheckVerdict = "within_timer" | "before_hold" | "after_timer" | "not_visible";
export type ReceiverMatchVerdict = "match" | "mismatch" | "not_configured" | "unavailable";

export interface Payment {
  id: string;
  booking_id: string;
  proof_url: string | null;
  proof_hash: string | null;
  expected_amount: number | null;
  ocr_amount: number | null;
  ocr_ref: string | null;
  ocr_verdict: OcrVerdict | null;
  ocr_confidence: number | null;
  // Section 32 Part 7
  ocr_payer_name: string | null;
  ocr_bank: string | null;
  ocr_receiver: string | null;
  ocr_flags: string[] | null;
  name_match_verdict: NameMatchVerdict | null;
  time_check_verdict: TimeCheckVerdict | null;
  receiver_match_verdict: ReceiverMatchVerdict | null;
  is_duplicate: boolean;
  duplicate_of: string | null;
  review_verdict: ReviewVerdict | null;
  rejection_reason: string | null;
  reviewed_at: string | null;
  auto_approved: boolean;
  created_at: string;
}

/** One plain-language check for the owner's approval card -- `text` is a
 * ready-made sentence from the backend; never re-derive or hand-roll it
 * client-side (same "hand it a ready-made string" convention this codebase
 * already uses for money/time in the AI chat). */
export interface PaymentCheck {
  verdict: "match" | "mismatch" | "warning" | "not_available" | "not_configured";
  text: string;
}

/** Section 32 Part 7 -- the five plain-language checks shown on the
 * owner's approval card. Purely informational: nothing here means the
 * payment was rejected, only that the owner should look closer. */
export interface PaymentChecks {
  name: PaymentCheck;
  amount: PaymentCheck;
  balance_text: string;
  time: PaymentCheck;
  bank: PaymentCheck;
  duplicate: PaymentCheck;
}

export interface PaymentProofOut {
  payment: Payment;
  booking: Booking;
}

export interface ProofUrlOut {
  url: string;
  expires_in: number;
}

File: court-booking-backend/app/schemas/payment.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.booking import BookingOut


class PaymentRejectIn(BaseModel):
    reason: str


class PaymentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    booking_id: uuid.UUID
    proof_url: str | None = None
    proof_hash: str | None
    expected_amount: float | None = None
    ocr_amount: float | None
    ocr_ref: str | None
    ocr_verdict: str | None
    ocr_confidence: float | None
    # Section 32 Part 7
    ocr_payer_name: str | None = None
    ocr_bank: str | None = None
    ocr_receiver: str | None = None
    ocr_flags: list[str] | None = None
    name_match_verdict: str | None = None
    time_check_verdict: str | None = None
    receiver_match_verdict: str | None = None
    is_duplicate: bool
    duplicate_of: uuid.UUID | None
    review_verdict: str | None
    rejection_reason: str | None
    reviewed_at: datetime | None
    auto_approved: bool
    created_at: datetime


class PaymentCheckOut(BaseModel):
    """One plain-language check for the owner's approval card -- `verdict`
    drives the tick/warning/cross icon, `text` is the ready-made sentence to
    display (never hand-rolled client-side, same "hand it a ready-made
    string" convention this codebase already uses for money/time in the AI
    chat)."""

    verdict: str  # "match" | "mismatch" | "warning" | "not_available" | "not_configured"
    text: str


class PaymentChecksOut(BaseModel):
    """Section 32 Part 7: the five plain-language checks shown on the
    owner's approval card. Auto-approve requires name/amount/time/receiver
    all "match" (or receiver "not_configured", see build_payment_checks) and
    no duplicate -- but this object itself is purely informational; nothing
    here auto-rejects anything, the owner always makes the final call."""

    name: PaymentCheckOut
    amount: PaymentCheckOut
    balance_text: str
    time: PaymentCheckOut
    bank: PaymentCheckOut
    duplicate: PaymentCheckOut


class PaymentSubmitResponse(BaseModel):
    payment: PaymentOut
    booking: BookingOut


class ProofUrlOut(BaseModel):
    url: str
    expires_in: int
